File: JavaScript/keygen.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Function to generate RSA key pair and save to files
function generateAndSaveKeys() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
        modulusLength: 2048,  // Standard key size for RSA
    });

    // Convert keys to PEM format (string)
    const publicKeyPem = publicKey.export({ type: 'spki', format: 'pem' });
    const privateKeyPem = privateKey.export({ type: 'pkcs8', format: 'pem' });

    fs.writeFileSync(path.join(__dirname, 'publicKey.pem'), publicKeyPem);
    fs.writeFileSync(path.join(__dirname, 'privateKey.pem'), privateKeyPem);

    console.log('RSA key pair generated and saved.');
}

// Function to load the public key from file
function loadPublicKey() {
    try {
        return fs.readFileSync(path.join(__dirname, 'publicKey.pem'), { encoding: 'utf8' });
    } catch (error) {
        console.error('Error loading the public key:', error.message);
        return null;
    }
}

// Function to load the private key from file
function loadPrivateKey() {
    try {
        return fs.readFileSync(path.join(__dirname, 'privateKey.pem'), { encoding: 'utf8' });
    } catch (error) {
        console.error('Error loading the private key:', error.message);
        return null;
    }
}

function generateAESKey() {
  return crypto.randomBytes(32); // AES-256 key
}

function generateDHKeys() {
    const dh = crypto.createDiffieHellmanGroup('modp14');
    const publicKey = dh.generateKeys();
    const privateKey = dh.getPrivateKey();

    return { publicKey, privateKey, dh };
}

function getPublicKeyFingerprint() {
    try {
        const publicKey = loadPublicKey();
        const hash = crypto.createHash('sha256').update(publicKey).digest('hex');
        return hash; // Returns the fingerprint
    } catch (error) {
        console.error('Error generating public key fingerprint:', error);
        return null;
    }
}

// Function to regenerate and save new keys
function regenerateAndSaveKeys() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', { modulusLength: 2048 });
    const publicKeyPem = publicKey.export({ type: 'spki', format: 'pem' });
    const privateKeyPem = privateKey.export({ type: 'pkcs8', format: 'pem' });
    fs.writeFileSync(path.join(__dirname, 'publicKey.pem'), publicKeyPem);
    fs.writeFileSync(path.join(__dirname, 'privateKey.pem'), privateKeyPem);
    console.log('New RSA key pair generated and saved.');
}

module.exports = { generateAndSaveKeys, loadPublicKey, loadPrivateKey, generateAESKey, generateDHKeys, getPublicKeyFingerprint, regenerateAndSaveKeys};

File: Python/keygen.py
from cryptography.hazmat.primitives.asymmetric import rsa, dh
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.serialization import load_pem_public_key, load_pem_private_key
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend
import os

def generate_and_save_keys():
    try:
        private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
        public_key = private_key.public_key()

        # Save the private key
        with open(os.path.join(os.path.dirname(__file__), 'privateKey.pem'), 'wb') as f:
            f.write(private_key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption()
            ))

        # Save the public key
        with open(os.path.join(os.path.dirname(__file__), 'publicKey.pem'), 'wb') as f:
            f.write(public_key.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            ))

        print('RSA key pair generated and saved successfully.')
    except Exception as error:
        print(f'Failed to generate or save RSA key pair: {error}')

def load_public_key():
    try:
        with open(os.path.join(os.path.dirname(__file__), 'publicKey.pem'), 'rb') as f:
            return load_pem_public_key(f.read())
    except Exception as error:
        print(f'Failed to load public key: {error}')
        return None

def load_private_key():
    try:
        with open(os.path.join(os.path.dirname(__file__), 'privateKey.pem'), 'rb') as f:
            return load_pem_private_key(f.read(), None)
    except Exception as error:
        print(f'Failed to load private key: {error}')
        return None

def generate_aes_key():
    return os.urandom(32)

def generate_dh_keys():
    try:
        parameters = dh.generate_parameters(generator=2, key_size=1024, backend=default_backend())
        private_key = parameters.generate_private_key()
        public_key = private_key.public_key()
        return {'public_key': public_key, 'private_key': private_key}
    except Exception as error:
        print(f'Failed to generate DH keys: {error}')
        return None

def get_public_key_fingerprint(public_key):
    try:
        # Assuming public_key is already loaded and is of type RSAPublicKey
        pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        digest = hashes.Hash(hashes.SHA256())
        digest.update(pem)
        return digest.finalize().hex()
    except Exception as error:
        print(f'Failed to generate public key fingerprint: {error}')
        return None

def regenerate_and_save_keys():
    try:
        generate_and_save_keys()
        print('Keys have been regenerated and saved.')
    except Exception as error:
        print(f'Failed to regenerate keys: {error}')
